src/main/java/com/amazon/ion/impl/macro/PooledExpressionFactory.java

```diff
@@ -0,0 +1,380 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.impl.macro;
+
+import com.amazon.ion.IonType;
+import com.amazon.ion.SymbolToken;
+import com.amazon.ion.Timestamp;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A factory for {@link Expression} instances. Avoids repetitive allocations by pooling instances. {@link #clear()}
+ * returns all instances to the pool. Note: this should only be used when the lifetime of the provided instances
+ * is known, and that lifetime must be the same for all instances provided between calls to {@link #clear()}.
+ */
+public class PooledExpressionFactory {
+
+    private static final int POOL_SIZE = 32;
```

Comment:
This confused me (briefly) further down the page. Might be worth renaming to `INITIAL_POOL_SIZE` or something like that.